packages/server/graphql/intranetSchema/mutations/disconnectSocket.ts

```diff
@@ -1,50 +1,50 @@
 import {SubscriptionChannel} from 'parabol-client/types/constEnums'
-import getRethink from '../../../database/rethinkDriver'
 import db from '../../../db'
 import {getUserId} from '../../../utils/authorization'
 import publish from '../../../utils/publish'
 import segmentIo from '../../../utils/segmentIo'
 import {GQLContext} from '../../graphql'
 import DisconnectSocketPayload from '../../types/DisconnectSocketPayload'
-
+import getRedis from '../../../utils/getRedis'
+import {UserPresence} from './connectSocket'
 export default {
   name: 'DisconnectSocket',
   description: 'a server-side mutation called when a client disconnects',
   type: DisconnectSocketPayload,
   resolve: async (_source, _args, {authToken, socketId}: GQLContext) => {
     // Note: no server secret means a client could call this themselves & appear disconnected when they aren't!
-    const r = await getRethink()
+    const redis = getRedis()
 
     // AUTH
     if (!socketId) return undefined
     const userId = getUserId(authToken)
 
     // RESOLUTION
-    const reqlUpdater = (user) => ({
-      lastSeenAtURL: r.branch(
-        user('connectedSockets')
-          .count()
-          .eq(0),
-        null,
-        user('lastSeenAtURL')
-      ),
-      connectedSockets: user('connectedSockets')
-        .default([])
-        .difference([socketId])
-    })
-    const user = await db.write('User', userId, reqlUpdater)
-    const {tms, connectedSockets} = user
-    const data = {user}
-    if (connectedSockets.length === 0) {
-      // If that was the last socket, tell everyone they went offline
-      const listeningUserIds = await r
-        .table('TeamMember')
-        .getAll(r.args(tms), {index: 'teamId'})
-        .filter({isNotRemoved: true})('userId')
-        .distinct()
-        .run()
+    const user = await db.read('User', userId)
+    const {tms} = user
+    const userPresence = await redis.lrange(`presence:${userId}`, 0, -1)
+    const parsedUserPresence = userPresence.map((socket) => JSON.parse(socket)) as UserPresence[]
+    const disconnectingSocket = parsedUserPresence.find((socket) => socket.socketId === socketId)
+    if (!disconnectingSocket) {
+      throw new Error('Called disconnect without a valid socket')
+    }
+    await redis.lrem(`presence:${userId}`, 0, JSON.stringify(disconnectingSocket))
+
+    // If this is the last socket, tell everyone they're offline
+    if (userPresence.length === 1) {
+      const listeningUserIds = new Set()
+      await redis.srem('onlineUserIds', userId)
+      for (const teamId of tms) {
+        const teamMembers = await redis.smembers(`team:${teamId}`)
```

Comment:
multi.exec here, too.
could refactor into a helper function like `getListeningUserIds(tms, commands)` where it calls a couple redis commands and returns an array of listeningUserIds